src/system/SystemPacketBuffer.h

```diff
@@ -230,3 +243,86 @@ inline uint16_t PacketBuffer::AllocSize() const
 
 } // namespace System
 } // namespace chip
+
+namespace chip {
+namespace System {
+
+/// Tracks ownership of a System::PacketBuffer.
+class DLL_EXPORT PacketBufferHandle
+{
+public:
+    PacketBufferHandle() : mBuffer(nullptr) {}
+    PacketBufferHandle(decltype(nullptr)) : mBuffer(nullptr) {}
+
+    // The caller's ownership is transferred to this.
+    explicit PacketBufferHandle(PacketBuffer * buffer) : mBuffer(buffer) {}
+
+    PacketBufferHandle(PacketBufferHandle && aOther)
+    {
+        mBuffer        = aOther.mBuffer;
+        aOther.mBuffer = nullptr;
+    }
+
+    ~PacketBufferHandle() { Adopt(nullptr); }
+    void Free() { Adopt(nullptr); }
+
+    PacketBufferHandle & operator=(PacketBufferHandle && aOther)
+    {
+        if (mBuffer)
+        {
+            PacketBuffer::Free(mBuffer);
+        }
+        mBuffer        = aOther.mBuffer;
+        aOther.mBuffer = nullptr;
+        return *this;
+    }
+    PacketBufferHandle & operator=(decltype(nullptr))
+    {
+        Adopt(nullptr);
+        return *this;
+    }
+
+    PacketBufferHandle Clone() const
+    {
+        mBuffer->AddRef();
+        return PacketBufferHandle(mBuffer);
+    }
+
+    PacketBuffer * operator->() { return mBuffer; }
+    PacketBuffer & operator*() { return *mBuffer; }
+
+    const PacketBuffer * operator->() const { return mBuffer; }
+    const PacketBuffer & operator*() const { return *mBuffer; }
+
+    // The caller's ownership is transferred to this.
+    void Adopt(PacketBuffer * buffer)
+    {
+        if (mBuffer != nullptr)
+        {
+            PacketBuffer::Free(mBuffer);
+        }
+        mBuffer = buffer;
+    }
+
+    // The PacketBufferHandle's ownership is transferred to the caller.
+    CHECK_RETURN_VALUE PacketBuffer * Release()
+    {
+        PacketBuffer * buffer = mBuffer;
+        mBuffer               = nullptr;
+        return buffer;
+    }
+
+    // The caller has access but no ownership.
+    PacketBuffer * Get_NoRelease() { return mBuffer; }
```

Comment:
The lack of parallelism in `Release` and `Get_NoRelease` seems a bit strange. I'd advocate for either `Release` and `Get` or `Get_Release` and `Get_NoRelease`.

Reply:
Agree they are odd. My preference would be to not have 'Get' at all if possible as it gives a raw pointer without as clear of a lifetime. This is why the original AutoFree was meant as a marker of 'this get is odd'.j

Not sure if this applies anymore. Maybe some TODO to drop this Get and replace with casts to appropriate types (e.g. when using LWIP) would be best. Otherwise we can make the naming uniform.

Reply:
Yes, this is intended to be temporary, since eventually nothing within CHIP should take a raw PacketBuffer. There may be some remaining use involving LWIP (I don't know yet whether all such uses take ownership) in which case I was intending to add an LWIP-specific access with some clear name.

Reply:
As requested below, I'll rename this `Get_ForNow` and create a permanent `Get` if/when the need is clear.